scripts/testing/e2e-runner

```diff
@@ -0,0 +1,348 @@
+
+    for var in ${!http*} ${!HTTP*} ${!no_*} ${!NO_*}; do
+        case $var in
+            *_proxy|*_PROXY)
+                info "exporting proxy envrionment variable $var (with value ${!var})..."
+                export $var
+                ;;
+        esac
+    done
+}
+
+prepare_worktree() {
+    local worktree_dir="$(dirname $CLONED_REPO)/worktree"
+
+    WORKTREE="$worktree_dir/$RESULT_NAME"
+    case "$SKIP_WORKTREE" in
+        1|true|yes|y)
+            WORKTREE="$(realpath $WORKTREE)"
+            if ! cd $WORKTREE; then
+                fatal "failed to cd into worktree $WORKTREE..."
+            fi
+            return 0
+            ;;
+    esac
+
+    if [ ! -e "$CLONED_REPO" ]; then
+        if ! mkdir -p $CLONED_REPO; then
+            fatal "failed to create directory $CLONED_REPO"
+        fi
+        cd $CLONED_REPO
+        if ! git clone $REMOTE_REPO .; then
+            fatal "failed to clone $REMOTE_REPO to $CLONED_REPO"
+        fi
+    else
+        cd $CLONED_REPO
+        if ! git remote -v | grep origin | grep -q $REMOTE_REPO; then
+            fatal "existing clone $CLONED_REPO has other origin than $REMOTE_REPO"
+        fi
+        if ! git fetch origin; then
+            fatal "failed to fetch $REMOTE_REPO"
+        fi
+    fi
+
+    if ! mkdir -p "$worktree_dir"; then
+        fatal "failed to create worktree root $worktree_dir"
+    fi
+
+    WORKTREE="$(realpath $WORKTREE)"
+    info "setting up git worktree $WORKTREE..."
+    if ! git worktree add $WORKTREE origin/$TEST_BRANCH; then
+        fatal "failed to add git worktree $WORKTREE"
+    fi
+
+    if [ -x "$WORKTREE/scripts/testing/e2e-runner" ]; then
+        exec "$WORKTREE/scripts/testing/e2e-runner" \
+             --skip-worktree \
+             ${PROXIES_SRC:+--source-proxies }${PROXIES_SRC:-} \
+             --remote $REMOTE_REPO \
+             --branch $TEST_BRANCH \
+             --local $CLONED_REPO \
+             --results $RESULT_ROOT \
+             --name $RESULT_NAME \
+             $E2E_TESTS
+    else
+        cd $WORKTREE
+    fi
+}
+
+prepare_result_dir() {
+    local dir="$RESULT_DIR"
+    local log="$dir/${SELF##*/}.log"
+
+    info "creating result dir $dir..."
+    if ! mkdir -p "$dir"; then
+        fatal "failed to create $dir"
+    fi
+
+    if ! touch "$log"; then
+        fatal "failed to create log file $log"
+    fi
+    LOG="$log"
+
+    info "Redirecting output to log file $LOG..."
+    if ! exec &>"$LOG"; then
+        fatal "log redirection failed"
+    fi
+}
+
+run_e2e_tests() {
+    local tests
+
+    info "running e2e tests ${E2E_TESTS:-with the default test set}..."
+    cd $WORKTREE
+    if [ -n "$E2E_TESTS" ]; then
+        make E2E_TESTS="$E2E_TESTS" e2e-tests
+    else
+        make e2e-tests
+    fi
+}
+
+cleanup_test_vms() {
+    local src
+
+    info "cleaning up test VMs..."
+    cd $WORKTREE
+    for src in test/e2e/n[0-9]*c[0-9]*-c*; do
+        ([ -d "$src" ] || exit 0; cd $src; vagrant destroy --force)
```

Comment:
let's not destroy if `cd` fails... parenthesis could be simplified into
```
(cd "$src" && vagrant destroy --false)
```

Reply:
Updated.